lib/at_ex/gateway/Sms/bulk.ex

```diff
@@ -23,7 +23,7 @@ defmodule AtEx.Gateway.Sms.Bulk do
       |> Map.put(:username, username)
 
     with {:ok, %{status: 201} = res} <- post("/messaging", params) do
-      {:ok, Jason.decode!(res.body)}
+      process_result(res.body)
```

Comment:
Hey @TraceyOnim looking good, was thinking of having this refactor pulled in, can you first confirm one thing for instance we have this code 
`    
{:ok, Jason.decode!(res.body)}
    else
      {:ok, val} ->
        {:error, %{status: val.status, message: val.body}}

      {:error, message} ->
        {:error, message}
    end `
which seems common among all our endpoints can we see if the `process_result` can include all of it since at the moment it only has this code:
`
with {:ok, res} <- Jason.decode(result) do
          {:ok, res}
        else
          {:error, val} ->
            {:error, val}
        end
`
Was thinking of it to be something like this:

`
  def process_result(result) do
        with {:ok, res} <- Jason.decode(result) do
          {:ok, res}
        else
          {:ok, val} ->
            {:error, %{status: val.status, message: val.body}}

          {:error, message} ->
            {:error, message}
        end
      end
`
Thus we can have the `send_sms`to be like this: 
`
  def send_sms(attrs) do
    username = Application.get_env(:at_ex, :username)

    params =
      attrs
      |> Map.put(:username, username)

    with {:ok, %{status: 201} = res} <- post("/messaging", params) do
      process_result(res.body)
    end
  end
`

Reply:
@kamalogudah this looks fine, actually from what we discussed form the previous session, for the responses where there's an actual error within the response from Africastalking

Reply:
@kamalogudah thats actually a better refactor for that function.  I will look into it and ensure nothing is broken

Reply:
@kamalogudah  I  don't think the suggested refactor will work as expected. `process_result` takes in  the result  which is the body of the response then decodes it .  From your suggestion it only means that  result  is processed only if the response we get has `status 200`
``` elixir
{:ok,
 %Tesla.Env{
   __client__: %Tesla.Client{adapter: nil, fun: nil, post: [], pre: []},
   __module__: AtEx.Gateway.Sms.Bulk,
   body: "{\"SMSMessageData\":{\"Message\":\"Sent to 1/1 Total Cost: KES 0.8000\",\"Recipients\":[{\"statusCode\":101,\"number\":\"+254718539545\",\"cost\":\"KES 0.8000\",\"status\":\"Success\",\"messageId\":\"ATXid_a04bcfa294cdb523d9a9f70632c0c167\"}]}}",
   headers: [
     {"date", "Sat, 20 Jun 2020 12:04:12 GMT"},
     {"server", "akka-http/10.0.5"},
     {"content-length", "218"},
     {"content-type", "application/json"}
   ],
   method: :post,
   opts: [],
   query: [],
   status: 201,
   url: "https://api.sandbox.africastalking.com/version1/messaging"
 }}

```

Reply:
@kamalogudah  I have worked on it, kindly check it out

Reply:
@TraceyOnim tried it locally and it was breaking stuff, we can take a second look later.

Reply:
Looking good, just few other comments and I think we will be good to go.